controller/workitem.go

```diff
@@ -561,6 +563,199 @@ func getVersion(version interface{}) (int, error) {
 	return -1, nil
 }
 
+// ConvertWorkItemsToCSV is responsible for converting given []WorkItem model object into a
+// []string object containing a set of CSV formatted data lines and a header line with labels.
+// This methods combines all CSV data of all WITs into a single CSV
+func ConvertWorkItemsToCSV(ctx context.Context, app application.Application, wits []workitem.WorkItemType, wis []workitem.WorkItem) (string, error) {
+	// uuidStringCache stores the UUID cache for ID resolvings of the CSV conversion
+	uuidStringCache := map[string]string{}
+	if len(wits) != len(wis) {
+		return "", errs.Errorf("length mismatch of work items (%d) and work item types (%d)", len(wis), len(wits))
+	}
+	// columnMap holds the mapping from column key to column index.
+	// This is there to not be required to iterate over the label line
+	// on every field. Makes the algorithm more efficient on the cost of
+	// a bit more memory.
+	columnMap := make(map[string]int)
+	// csvGrid holds the final CSV in non-serialized form.
+	csvGrid := [][]string{}
+	// add an empty label line by default.
+	csvGrid = append(csvGrid, []string{})
+	for i := 0; i < len(wis); i++ {
+		// iterate over the WIs, converting them one by one.
+		fieldLabels, _, fieldValues, err := ConvertWorkItemToStringValue(ctx, app, &uuidStringCache, wits[i], wis[i])
+		if err != nil {
+			return "", errs.Wrapf(err, "failed to convert work item to CSV: %s", wis[i].ID)
+		}
+		// create a new line, yet without potential new columns.
+		csvLine := make([]string, len(csvGrid[0]))
+		// add new new (empty) line to the grid.
+		csvGrid = append(csvGrid, csvLine)
+		// now, mangle up the column configuration
+		for j := 0; j < len(fieldLabels); j++ {
+			// iterate over the field keys, mapping them to the resulting grid
+			// the CSV key is unique and contains the WIT's name and the fieldLabel
+			// there might be collisions due to WITs names and/or fieldKeys are not
+			// required to be unique, but for the usecase, if there is an equally
+			// named WIT/fieldKey in a template, it is highly probable that this
+			// is intended to go into the same column.
+			csvFieldKey := fieldLabels[j]
+			// now check if the current key is a new key.
+			_, ok := columnMap[csvFieldKey]
+			if !ok {
+				// this is a new key, create a new columnMap entry.
+				columnMap[csvFieldKey] = len(csvGrid[0])
+				// add the new header entry.
+				csvGrid[0] = append(csvGrid[0], csvFieldKey)
+				// append empty values to all existing "lines" for the new column.
+				for k := 1; k < len(csvGrid); k++ {
```

Comment:
We could've created the `columnMap` before adding workitems to the CSV.
Right now we have 
```
1. For each fieldLabel (say n fieldLabelss)
    - Check if field is in columnMap. If not, add new column and append "" to all the existing lines/workitems (say m existing lines)
```
So the worst case (where all the workitems have different columns) time complexity is O(n * m).

I would've instead done the following
```
1. Create ColumnMap using fieldLabels (say n field labels)
2. Loop through the field labels and add build each row of CSV (since we do a map lookup via columnMap[csvFieldKey], we can add "" if they key isn't present) (say n field labels)
```
The time complexity for the above algorithm would be `O(n + n)` which is `O(n)`.
This might not have serious performace improvement but a lot of small improvements can have significant effect on the outcome :)

What do you think @michaelkleinhenz @kwk ?

Reply:
@jarifibrahim are you suggesting, among other things, to get a list of all field labels upfront? If so that means we have to iterate over all WIs in order to get a distinct map of all WITs involved, right?

Reply:
> @jarifibrahim are you suggesting, among other things, to get a list of all field labels upfront? If so that means we have to iterate over all WIs in order to get a distinct map of all WITs involved, right?

Correct. I'm suggesting we build the `list of columns` before we start adding values to the CSV table.

Reply:
Fixed 7ae9140.

Reply:
Is now building the column config upfront.

Reply:
Thank you :)